### PR TITLE
OOP exercise s ready for review

### PR DESCRIPTION
**Pull request by m-ruslan**
### OOP exercise

[Demo](https://m-ruslan.github.io/a-tiny-JS-world/) |
[Code base](https://github.com/m-ruslan/a-tiny-JS-world)

The code is submitted in a dedicated feature branch.
Please, review.

### CODE REVIEW

**Review by OleksiyRudenko — CHANGES_REQUESTED** [review]
@m-ruslan nice job!
A note: functions/methods are made a part of prototype to save memory and computation power during execution/objects construction. Properties storing primitive values like strings and numbers take memory anuway and thus are fine to be initialized via `this` in constructor function.

Note 2: if properties for hands and legs are strings then it would be difficult to count limbs if we wanted. Besides string values are somewhat tautological and also not intenationalazible.

Major point:
Please refactor your code employing es6 classes.
[TJSW project](https://github.com/OleksiyRudenko/a-tiny-JS-world) offers materials on es6 classes, OOP and SOLID you may find useful.

A checklist for your convenience:

Minimal requirements to meet:
 * [ ] Implement a base class to inherit from
 * [ ] Employ default parameters
 * [ ] Each species is represented with its own class
 * [ ] No need to specify species at instantiation
 * [ ] Classes for species that do not have hands by natural design
       do not consequently have `hands` or any equivalent property
 * [ ] All inhabitants are stored in a container (array or object)
 * [ ] JS native features are intensively employed (`const`, `let`, `Array.map|join|forEach|...`, etc)
 * [ ] [OOP, SOLID and DRY](https://github.com/OleksiyRudenko/a-tiny-JS-world/blob/master/README.md#learn-on-your-own)
       principles are intensively employed
 
Optional level up (not required to implement):
 * [ ] Friends list is a list of objects refs rather than names (strings)
 * [ ] Cat-woman class is built employing composition rather than inheritance only

**Comment by m-ruslan** [comment]
@OleksiyRudenko thank you very much)
I will refactor my code to make it fit to checklist you added.

Maybe I've missed this checklist in the task description, but if it is not there then probably it would be helpful for other students if this checklist was specified in the task description)

**Comment by OleksiyRudenko** [comment]
The point of the course is to make students' life easier by themselves learning from the materials, analysing etc.
Basically the requirements in the checklist are distillation of the materials referred to [here](https://github.com/OleksiyRudenko/a-tiny-JS-world#leveling-your-skills-up). These materials are already a part of the course.

**Comment by m-ruslan** [comment]
>Unfortunately, you won't always have simple checklists to complete a work task.

I totally agree with this sentence
@OleksiyRudenko, thank you
I will come back to you soon with the refactored code)

**Comment by m-ruslan** [comment]
@OleksiyRudenko it seems to me that I've fixed all issues to meet minimal requirements.
Could you please review?

Also I've employed composition building `Cat-woman` class and `Cat` class.  `Man`, `Woman` and `Dog` classes built using ES6 classes.

But I haven't figured out how to implement this:
> Friends list is a list of objects refs rather than names (strings)

When in `dog` object creation I refer to `woman` object (that is assigned lower) the error occurs because `woman` is not assigned yet.

**Comment by OleksiyRudenko** [comment]
> When in `dog` object creation I refer to `woman` object (that is assigned lower) the error occurs because `woman` is not assigned yet.

Didn't look at the code yet. The problem above is quite solvable.
E.g.
```
const dog = new Dog();
const man = new Human(..., [dog]);
conts woman = new Human(..., [dog, man]);
man.addFriends([woman]);
dog.addFriends([woman, man]);

/// or
const friendsList = [
 { target: dog, friends: [man, woman], },
 { target: man, friends: [dog, woman], },
 { target: woman, friends: [dog, man], },
].forEach(({target, friends}) => target.addFriends(friends));

// addFriend() might also work both ways - adding a friend makes current object a friend to the 'to-be-a-friend'

```

**Comment by m-ruslan** [comment]
@OleksiyRudenko thank you for the review. Could you please check it one more time, I've tried to fix the issues you wrote about.

File: submissions/m-ruslan/oop-exercise/index.js [hunk 1]
```diff
@@ -0,0 +1,192 @@
...
+// Define your objects here
+
+//**ES6 classes and sub-classes
+class Inhabitant {
+  constructor(name, legs, gender, propNamesArr) {
+    this.name = name;
+    this.gender = gender;
+    this.legs = legs;
+    this.friendTo = [];
+    this.propNamesArr = propNamesArr;
```
Comment (OleksiyRudenko) [inline submissions/m-ruslan/oop-exercise/index.js:18]: I see the problem you're trying to resolve by holding `propNamesArr`.
Not illegal, yet not really elegant.

**Review by OleksiyRudenko — APPROVED** [review]
@m-ruslan a really great job!
Mating ~a hedgehog with a snake~ es6 classes with compositional pattern is a challenge. This naturally leads to not really elegant solutions, for this approaches are generally incompatible.
You managed it. Grats!

Although the major learning here: try it, succeed with it, and... never do this again :)

**Comment by m-ruslan** [comment]
@OleksiyRudenko thank you very much)

**Pull request merged**